runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala

```diff
@@ -33,38 +33,42 @@ import akka.util.OptionVal
 @InternalApi
 final class ScalaUnaryRequestBuilder[I, O](
   descriptor: MethodDescriptor[I, O],
-  channel: Channel,
+  channel: Future[ManagedChannel],
   defaultOptions: CallOptions,
   settings: GrpcClientSettings,
-  val headers: MetadataImpl)
+  val headers: MetadataImpl)(implicit ec: ExecutionContext)
   extends akka.grpc.scaladsl.SingleResponseRequestBuilder[I, O]
   with MetadataOperations[ScalaUnaryRequestBuilder[I, O]] {
 
   // aux constructor for defaults
-  def this(descriptor: MethodDescriptor[I, O], channel: Channel, defaultOptions: CallOptions, settings: GrpcClientSettings) =
+  def this(descriptor: MethodDescriptor[I, O], channel: Future[ManagedChannel], defaultOptions: CallOptions, settings: GrpcClientSettings)(implicit ec: ExecutionContext) =
     this(descriptor, channel, defaultOptions, settings, MetadataImpl.empty)
 
   private def callOptionsWithDeadline(): CallOptions =
     NettyClientUtils.callOptionsWithDeadline(defaultOptions, settings)
 
   override def invoke(request: I): Future[O] = {
-    val listener = new UnaryCallAdapter[O]
-    val call = channel.newCall(descriptor, callOptionsWithDeadline())
-    call.start(listener, headers.toGoogleGrpcMetadata())
-    call.sendMessage(request)
-    call.halfClose()
-    call.request(2)
-    listener.future
+    channel.flatMap(c => {
+      val listener = new UnaryCallAdapter[O]
+      val call = c.newCall(descriptor, callOptionsWithDeadline())
+      call.start(listener, headers.toGoogleGrpcMetadata())
+      call.sendMessage(request)
+      call.halfClose()
+      call.request(2)
+      listener.future
+    })
```

Comment:
I wonder if we should optimize this right away for the case where the channel future is already completed (which should be most of the invocations) and avoid one extra dispatch on the ec?

Reply:
👍